// @cpmscf/mdlib/out/dita/upload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DitaUploader = void 0;
const tslib_1 = require("tslib");
const https_1 = require("https");
const path_1 = require("path");
const FormData = require("form-data");
const fs_1 = require("fs");
function checkOptions(opts) {
    if (opts.upload === undefined || opts.upload.auth === undefined)
        throw new TypeError("Missing required 'upload.auth' value for upload.");
    if (opts.dita === undefined || opts.dita.outputPath === undefined)
        throw new TypeError("Missing required 'dita.outputPath' value for upload.");
    if (opts.dita === undefined || opts.dita.deliverable === undefined ||
        opts.dita.deliverable.rootfile === undefined)
        throw new TypeError("Missing required 'dita.deliverable.rootFile' value for upload.");
    return opts;
}
class DitaUploader {
    upload(opts) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const options = checkOptions(opts);
            return new Promise((resolve, reject) => {
                const readStream = fs_1.createReadStream(path_1.join(opts.dita.outputPath, opts.dita.deliverable.rootfile));
                const form = new FormData();
                form.append("archive", readStream);
                const req = https_1.request({
                    protocol: "https:",
                    host: "uacp3.hana.ondemand.com",
                    port: "443",
                    path: "/http.svc/upload?from_github=true",
                    method: "POST",
                    headers: Object.assign(Object.assign({}, form.getHeaders()), { Authorization: `Basic ${options.upload.auth}` }),
                });
                form.pipe(req);
                req.on("error", reject);
                req.on("response", msg => {
                    if (msg.statusCode === undefined || msg.statusCode >= 300)
                        reject(new Error(`UACP upload request failed with status code ${msg.statusCode}`));
                    else
                        resolve(msg);
                });
            });
        });
    }
}
exports.DitaUploader = DitaUploader;
